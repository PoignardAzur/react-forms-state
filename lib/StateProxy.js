'use babel'
// @flow

import React from 'react';
import { fromJS } from "immutable";

export default function StateProxy(WrappedComponent) {
  class ProxiedState extends React.Component {
    render() {
      return (
        <WrappedComponent {...this.props}/>
      );
    }
  };

  ProxiedState.defaultProps = {
    _inputField: true,
    name: WrappedComponent.defaultProps.name,
  };

  return ProxiedState;
};
