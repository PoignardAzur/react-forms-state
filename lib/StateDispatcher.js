"use babel";
// @flow

import React from "react";
import { fromJS, is, List } from "immutable";
import StateProxy from "./StateProxy";
import { setValue, getValue } from "./StateValueHelpers";
import Rx from "rxjs";
import { INITIAL_CHANGE_EVENT_TYPE } from "./FormEvents";

export default function StateDispatcher(
  convertIn: (v: any, props: Props) => any = (v, props) => v,
  convertOut: (v: any, props: Props) => any = (v, props) => v,
) {
  return function(WrappedComponent: React$Component<any, any, any>) {
    class Dispatcher extends React.Component {
      subscription: Rx.Subscription;
      handlers: {[key: string]: {
        get: () => any,
        set: (value: any) => void,
      }};
      valueChangeObs: Rx.Observable;

      constructor(props) {
        super(props);

        this.handlers = {};
        this.valueChangeObs = props.valueChangeObs.map(e => {
          const convertedValue = convertIn(e.value, props);
          if (convertedValue === e.value)
            return e;
          else
            return Object.assign({}, e, {
              value: convertedValue,
              validation: e.validation,
              statePath: e.type === INITIAL_CHANGE_EVENT_TYPE
                ? undefined
                : props.statePath,
            });
        });
      }

      componentDidMount() {
        if (this.context.attachToController != null)
          this.context.attachToController(() => this.getUncontrolledState());
        this.subscription = this.valueChangeObs
          .filter(e => e.type == INITIAL_CHANGE_EVENT_TYPE)
          .do(
            e =>
              global.process &&
              global.process.env &&
              global.process.env.FORMALIZR_ENV === "DEBUG" &&
              console.log(
                `Dispatching uncontrolled values from ${this.props.statePath}`,
              ),
          )
          .subscribe({
            next: e => this.dispatchUncontrolledValues(e),
          });
      }

      componentWillUnmount() {
        if (this.context.attachToController != null)
          this.context.attachToController(null);
        this.subscription.unsubscribe();
      }

      dispatchUncontrolledValues(event) {
        Object.keys(this.handlers).forEach(statePath => {
          if (
            event.oldValue === undefined ||
            getValue(convertIn(event.value, this.props), statePath) !==
              getValue(convertIn(event.oldValue, this.props), statePath)
          )
            this.handlers[statePath].set(
              getValue(convertIn(event.value, this.props), statePath),
            );
        });
      }

      getChildContext() {
        return {
          valueChangeObs: this.valueChangeObs,
          rootValueChangeObs: this.props.rootValueChangeObs,
          onStateChange: (v, statePath, validation) => {
            const convertedInParentValue = convertIn(
              this.props.getValue(),
              this.props,
            );
            const completeValue = Array.isArray(convertedInParentValue) ===
              false
              ? Object.assign({}, convertedInParentValue, { [statePath]: v })
              : List().concat(convertedInParentValue).update(list => {
                  return list.update(
                    list.findIndex(item => item.id === statePath),
                    item => Object.assign({}, item, { value: v }),
                  );
                });
            const convertedValue = convertOut(completeValue, this.props);
            // console.log(
            //   "On change",
            //   completeValue,
            //   convertedValue,
            //   completeValue !== convertedValue,
            //   v,
            //   statePath,
            // );
            if (
              convertedValue !== completeValue ||
              typeof convertedValue != typeof this.props.getValue()
            )
              this.props.onChange(
                convertedValue,
                this.props.statePath,
                validation,
              );
            else
              this.props.onChange(
                v,
                `${this.props.statePath}${this.props.statePath == "" || statePath == "" ? "" : "."}${statePath}`,
                validation,
              );
          },
          attachToController: null,
          addHandler: ((statePath, getHandler, setHandler) => {
            this.handlers = Object.assign({}, this.handlers, {
              [statePath]: {
                get: getHandler,
                set: setHandler,
              },
            });
          }).bind(this),
        };
      }

      getUncontrolledState() {
        let value = Object.keys(this.handlers).reduce(
          (red, statePath) =>
            setValue(red, statePath, this.handlers[statePath].get()),
          {},
        );
        return convertOut(value, this.props);
      }

      render() {
        return <WrappedComponent {...this.props} />;
      }
    }

    Dispatcher.defaultProps = Object.assign({}, WrappedComponent.defaultProps, {
      onChange() {},
      value: {},
      valueChangeObs: Rx.Observable.never(),
    });

    Dispatcher.propTypes = {
      valueChangeObs: React.PropTypes.any,
      onChange: React.PropTypes.func,
    };

    Dispatcher.childContextTypes = {
      valueChangeObs: React.PropTypes.any,
      onStateChange: React.PropTypes.func,
      addHandler: React.PropTypes.func,
      attachToController: React.PropTypes.func,
      rootValueChangeObs: React.PropTypes.any,
    };

    Dispatcher.contextTypes = {
      attachToController: React.PropTypes.func,
    };


    return StateProxy(
      { root: true },
      {
        getValue: child => child.getUncontrolledState(),
        setValue: child => {},
      },
    )(Dispatcher);
  };
}

type Props = {
  valueChangeObs?: Rx.Observable;
  onChange?: (v: any) => void;
};
