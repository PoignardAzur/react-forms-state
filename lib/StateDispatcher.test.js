'use babel'
// @flow

import React from 'react';
import { shallow, mount } from 'enzyme';
import StateDispatcher from './StateDispatcher';
import StateProxy from "./StateProxy";

describe('StateDispatcher', () => {
  class A extends React.Component{render() { return (<span><strong>Ok</strong>Salut</span>) }};
  let InputField = StateProxy(A);
  class InputGroup extends React.Component {
    render() {
      return (
        <div>
          {this.props.children}
        </div>
      );
    }
  };
  let ProxiedGroup = StateProxy(InputGroup);

  it('should propagate value from an object', () => {
    class Form extends React.Component {
      render() {
        return (
          <div>
            <InputField name="firstname"/>
            <div>
              <p>
                <span>Hey</span>ok
              </p>
            </div>
            <div>
              <InputField name="lastname"/>
            </div>
            <InputField name="description"/>
          </div>
        );
      }
    };
    let ProxiedForm = StateDispatcher()(Form);
    let values = {
      firstname: "Alan",
      lastname: "Zanatta",
      description: "Good"
    };
    let subject = mount(<ProxiedForm value={values}/>);

    expect(subject.find(InputField).at(0).prop('value')).toBe('Alan');
    expect(subject.find(InputField).at(1).prop('value')).toBe('Zanatta');
    expect(subject.find(InputField).at(2).prop('value')).toBe('Good');
  });

  it('should aggregate values of an object for onChange', () => {
    class Form extends React.Component {
      render() {
        return (
          <div>
            <InputField name="firstname"/>
            <div>
              <InputField name="lastname"/>
            </div>
            <InputField name="description"/>
          </div>
        );
      }
    };
    let ProxiedForm = StateDispatcher()(Form);
    let values = {
      firstname: "Alan",
      lastname: "Zanatta",
      description: "Good"
    };
    let spy = jest.fn();
    let subject = mount(<ProxiedForm value={values} onChange={spy}/>);

    subject.find(InputField).at(1).prop('onChange')('Joe');

    expect(spy).toBeCalledWith({
      firstname: "Alan",
      lastname: "Joe",
      description: "Good"
    });
  });

  it('should propagate value from an array', () => {
    class Form extends React.Component {
      render() {
        return (
          <div>
            <InputField name="firstname"/>
            <div>
              <InputField name="lastname"/>
            </div>
            <InputField name="description"/>
          </div>
        );
      }
    };
    let ProxiedForm = StateDispatcher()(Form);
    let values = [{
      id: 'firstname',
      value: "Alan",
    }, {
      id: 'lastname',
      value: "Jon",
    }, {
      id: 'description',
      value: "Foo",
    }];
    let subject = mount(<ProxiedForm value={values}/>);

    expect(subject.find(InputField).at(0).prop('value')).toBe('Alan');
    expect(subject.find(InputField).at(1).prop('value')).toBe('Jon');
    expect(subject.find(InputField).at(2).prop('value')).toBe('Foo');
  });

  it('should aggregate values from an array for onChange', () => {
    class Form extends React.Component {
      render() {
        return (
          <div>
            <InputField name="firstname"/>
            <div>
              <InputField name="lastname"/>
            </div>
            <InputField name="description"/>
          </div>
        );
      }
    };
    let ProxiedForm = StateDispatcher()(Form);
    let values = [{
      id: 'firstname',
      value: 'Alan',
    }, {
      id: 'lastname',
      value: 'Zanatta',
    }, {
      id: 'description',
      value: 'Foo'
    }];
    let spy = jest.fn();
    let subject = mount(<ProxiedForm value={values} onChange={spy}/>);

    subject.find(InputField).at(1).prop('onChange')('Joe');

    expect(spy).toBeCalledWith([{
      id: 'firstname',
      value: 'Alan'
    }, {
      id: 'lastname',
      value: 'Joe'
    }, {
      id: 'description',
      value: "Foo",
    }]);
  });

  it('should compute and pass the state path', () => {
    class Form extends React.Component {
      render() {
        return (
          <div>
            <InputField name="firstname"/>
            <ProxiedGroup name="skills">
              <InputField name="tdd"/>
            </ProxiedGroup>
            <InputField name="description"/>
          </div>
        );
      }
    }
    let ProxiedForm = StateDispatcher()(Form);
    let values = {
      firstname: "Alan",
      description: "Good",
      skills: {
        tdd: false
      }
    };
    let subject = mount(<ProxiedForm value={values}/>);

    expect(subject.find(InputField).at(0).prop('statePath')).toBe('firstname');
    expect(subject.find(InputField).at(1).prop('statePath')).toBe('skills.tdd');
    expect(subject.find(InputField).at(2).prop('statePath')).toBe('description');
  });

  it('should stop propagation when another dispatcher is encountered', () => {
    class MiniForm extends React.Component {
      render() {
        return (
          <div>
            <InputField name="month"/>
            <InputField name="year"/>
          </div>
        );
      }
    }
    let ProxiedMiniForm = StateDispatcher(
      inValue => inValue.month == "1" ? {month: 'janvier', year: '2016'} : inValue,
      outvalue => outValue.month == "janvier" ? {month: "1", year: '2016'} : outvalue
    )(MiniForm);
    class Form extends React.Component {
      render() {
        return (
          <div>
            <InputField name="firstname"/>
            <ProxiedGroup name="skills">
              <InputField name="tdd"/>
            </ProxiedGroup>
            <ProxiedMiniForm name="date"/>
            <InputField name="description"/>
          </div>
        );
      }
    }
    let ProxiedForm = StateDispatcher()(Form);
    let values = {
      firstname: "Alan",
      description: "Good",
      skills: {
        tdd: false
      },
      date: {
        month: "1",
        year: "2016"
      }
    };
    let subject = mount(<ProxiedForm value={values}/>);

    expect(subject.find(ProxiedMiniForm).at(0).prop('value')).toEqual(values.date);
    expect(subject.find(ProxiedMiniForm).at(0).find(InputField).at(0).prop('value')).toBe('janvier');
    expect(subject.find(ProxiedMiniForm).at(0).find(InputField).at(1).prop('value')).toBe('2016');
  });
});
