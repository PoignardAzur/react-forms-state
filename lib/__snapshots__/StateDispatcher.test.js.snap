// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StateDispatcher Controlled should stop propagation when another dispatcher is encountered 1`] = `
Array [
  Object {
    "month": "2",
    "year": "17",
  },
  "birthday",
  true,
]
`;
